File: university_enrollment.c
/* 3. Create an university enrollment system with the following characteristics:
* 	The system has a login with a username and password.
* 	Upon logging in, a menu displays the available programs: Computer Science, Medicine, Marketing, and Arts.
* 	The user must input their first name, last name, and chosen program.
* 	Each program has only 5 available slots. The system will store the data of each registered user, and if it exceeds the limit, it should display a message indicating the program is unavailable.
* 	If login information is incorrect three times, the system should be locked.
* 	The user must choose a campus from three cities: London, Manchester, Liverpool.
* 	In London, there is 1 slot per program; in Manchester, there are 3 slots per program, and in Liverpool, there is 1 slot per program.
* 	If the user selects a program at a campus that has no available slots, the system should display the option to enroll in the program at another campus.
*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct user{
    char username[256];
    char password[256];
}user;

typedef struct courses{
    int slots_amount;
    char * students_name[256];
    char * students_last_name[256];
}courses;

typedef struct campus{
    struct courses slots[4];
}campus;

int main(){
    char * ptruser;
    char * ptrpassw;
    char * name;
    char * last_name;
    int user_counter=0;
    user users[256];
    int log_in_counter = 0;
    int counter;
    int menu1 = 1;
    int menu2;
    int menu3;
    int menu4;
    int option1;
    int course_option;
    int campus_option;
    int flag_slot_avialable;
    int flag_log_in;
    int slots_counter = 0;
    int amount_occupied_slots= 0;
    campus campus_slots[3];
    //Initializing the amount of slots in each campus
    int k;
    for(int i = 0; i < 3; i++){
        for(int j = 0; j < 4; j++){
            if(i == 1){
                k=3;
            }else{
                k=1;
            }
            campus_slots[i].slots[j].slots_amount=k;
        }
    }


    while(menu1==1){
        printf("University Enrollment System\n1-Create an Account\n2-Log in\n3-Exit\n");
        scanf("%i",&option1);
        switch (option1){
            //Creating an account
            case 1:
                ptruser = malloc(sizeof(char));
                ptrpassw = malloc(sizeof(char));
                printf("Enter your new credentials\nNew Username (no blank spaces):");
                scanf("%s",ptruser);
                printf("\nNew Password (no black spaces):");
                scanf("%s",ptrpassw);
                strcpy(users[user_counter].username,ptruser);
                strcpy(users[user_counter].password,ptrpassw);
                user_counter++;
                printf("Account created Successfuly\n\n");
                free(ptruser);
                free(ptrpassw);
                ptruser = NULL;
                ptrpassw = NULL;
                break;
            //Log In
            case 2:
                menu2 = 1;
                //Checking the credentials
                while(menu2 == 1){
                    ptruser = malloc(sizeof(char));
                    ptrpassw = malloc(sizeof(char));
                    printf("Enter your Credentials\nUsername:");
                    scanf("%s",ptruser);
                    printf("Enter your password:");
                    scanf("%s",ptrpassw);
                    flag_log_in = 0;
                    counter = 0;
                    while(counter < user_counter && menu2 == 1){
                        if(0 == strcmp(ptruser,users[counter].username) && 0 == strcmp(ptrpassw,users[counter].password)){
                            //Valid Credentials
                            menu3 = 1;
                            flag_slot_avialable = 0;
                            flag_log_in = 1;
                            slots_counter = 0;
                            while(menu3==1){
                                //Choosing a course
                                printf("Avaible Programs\n1-Computer Science\n2-Medicine\n3-Marketing\n4-Arts\n");
                                scanf("%i",&course_option);
                                for(int i = 0; i < 3;i++){
                                    //Checking available slots
                                    if(campus_slots[i].slots[course_option-1].slots_amount>0){
                                        flag_slot_avialable = 1;
                                        break;
                                    }
                                    if(flag_slot_avialable==1){
                                        break;
                                    }
                                }
                                //Is a slot avialable in the choosen course
                                if(flag_slot_avialable==1){
                                    menu4 = 1;
                                    name = malloc(sizeof(char));
                                    last_name = malloc(sizeof(char));
                                    //Asking the name and last name
                                    printf("There are slots available in the program.\n Please write your name:");
                                    scanf("%s",name);
                                    printf("Now please write your Last Name:");
                                    scanf("%s",last_name);
                                    //Choosing a campus
                                    while(menu4 == 1){
                                        printf("\nNow you need to choose a campus\n1-London\n2-Manchester\n3-Liverpool\n");
                                        scanf("%i",&campus_option);
                                        //Checking if the choosen campus have an avialable slot
                                        if(campus_slots[campus_option-1].slots[course_option-1].slots_amount > 0){
                                            //Saving the data
                                            printf("Congratulations, your enrollment have been succesful\n\n");
                                            campus_slots[campus_option-1].slots[course_option-1].students_name[slots_counter] = name;
                                            campus_slots[campus_option-1].slots[course_option-1].students_last_name[slots_counter]= last_name;
                                            menu3 = 0;
                                            menu2 = 0;
                                            slots_counter++;
                                            campus_slots[campus_option-1].slots[course_option-1].slots_amount --;
                                            break;
                                        }else{
                                            printf("There are not avialable slots in this campus, try another one\n\n");
                                        }
                                    }
                                    free(name);
                                    free(last_name);
                                    name = NULL;
                                    last_name = NULL;
                                }
                            }
                            if(flag_slot_avialable == 0){
                                printf("This course is unavailable, choose another \n\n");
                            }
                        }
                        counter++;
                    }
                    if(flag_log_in == 0){
                        printf("\nWrong Credentials\n\n");
                        log_in_counter++;
                        if(log_in_counter == 3){
                            printf("Too many attempts");
                            menu1= 0 ;
                            break;
                        }
                    }
                    free(ptruser);
                    free(ptrpassw);
                    ptruser = NULL;
                    ptrpassw = NULL;
                }

                break;
            case 3:
                menu1 = 0;
                break;
            default:
                menu1 = 0;
                break;
        }
    }

    return 0 ;
}

File: Ejercicios_lvl0/UniversityEnrollment.py
# Ejericio 3
# The system has a login with a username and password.
# Upon logging in, a menu displays the available programs: Computer Science, Medicine, Marketing, and Arts.
# The user must input their first name, last name, and chosen program.
# Each program has only 5 available slots. The system will store the data of each registered user, and if it exceeds the limit, it should display a message indicating the program is unavailable.
# If login information is incorrect three times, the system should be locked.
# The user must choose a campus from three cities: London, Manchester, Liverpool.
# In London, there is 1 slot per program; in Manchester, there are 3 slots per program, and in Liverpool, there is 1 slot per program.
# If the user selects a program at a campus that has no available slots, the system should display the option to enroll in the program at another campus.
from os import system

# Listas
london_1 = []
london_2 = []
london_3 = []
london_4 = []
manchester_1 = []
manchester_2 = []
manchester_3 = []
manchester_4 = []
liverpool_1 = []
liverpool_2 = []
liverpool_3 = []
liverpool_4 = []
computer_science = [london_1, manchester_1, liverpool_1]
medicina = [london_2, manchester_2, liverpool_2]
marketing = [london_3, manchester_3, liverpool_3]
artes = [london_4, manchester_4, liverpool_4]
estudiantes = [computer_science, medicina, marketing, artes]
lista_programas = ['computer science', 'medicina', 'marketing', 'artes']
lista_campus = ['london', 'manchester', 'liverpool']
valor_campus = [['london_1', 'manchester_1', 'liverpool_1'], ['london_2', 'manchester_2', 'liverpool_2'],
                ['london_3', 'manchester_3', 'liverpool_3'],
                ['london_4', 'manchester_4', 'liverpool_4']]
slots_disponible = {'london_1': 1, 'manchester_1': 3, 'liverpool_1': 1, 'london_2': 2, 'manchester_2': 3,
                    'liverpool_2': 1, 'london_3': 1, 'manchester_3': 3, 'liverpool_3': 1, 'london_4': 1,
                    'manchester_4': 3, 'liverpool_4': 1}


# Función inicio de sesión
def inicio_sesion():
    usuario = 'admin'
    password = 'python'
    intentos = 0
    while True:
        try:
            usuario_input = input('Ingrese un usuario: ')
            contrasena_input = input('Ingrese una contraseña: ')
            [usuario].index(usuario_input)
            [password].index(contrasena_input)

        except ValueError:
            intentos += 1
            system('cls')
            print('Ingresa las credenciales correctas.')

        else:
            system('cls')
            return True

        finally:
            if intentos == 3:
                system('cls')
                print('El sistema ha sido bloqueado debido a que ha hecho muchos intentos')
                return False


# Función para crear la ficha de un estudiante
def crear_estudiante():
    estudiante = {}
    while True:
        print('-' * 50)
        print('Ingrese los siguientes datos para crear su ficha de estudiante:')
        nombre = input('Escriba su nombre: ')
        segundo_nombre = input('Escriba su segundo nombre: ')
        try:
            for programa in lista_programas:
                print(programa.capitalize())
            programa = input('Escoge un programa de la lista anterior: ').lower()
            lista_programas.index(programa)
        except ValueError:
            system('cls')
            print('Error, el programa o campus elegido no existe')
        else:
            system('cls')
            estudiante['nombre'] = nombre
            estudiante['segundo nombre'] = segundo_nombre
            estudiante['programa'] = programa
            return estudiante


# Función para terminar la ficha de estudiante y elegir su campus
def elegir_campus(estudiante):
    textos_campus = ['london', 'manchester', 'liverpool']
    while True:
        print('-' * 50)
        for campus in textos_campus:
            print(campus.capitalize())
        try:
            campus = input('Escoge un campus de la lista anterior: ').lower()
            lista_campus.index(campus)
        except ValueError:
            system('cls')
            print('Error, el campus elegido no existe')
        else:
            estudiante_completo = verificar_disponibilidad(estudiante, campus)
            if estudiante_completo is not None:
                system('cls')
                print('Estudiante creado exitosamente')
                input('Presione cualquier tecla para continuar')
                system('cls')
                indice_campus = lista_campus.index(campus)
                indice_programa = lista_programas.index(estudiante['programa'])
                estudiantes[indice_programa][indice_campus].append(estudiante)
                return estudiante_completo
            else:
                textos_campus.remove(campus)
                continue


# Función para verificar si el campus elegido tiene slots disponible
def verificar_disponibilidad(estudiante, campus):
    programa_elegido = estudiante['programa']
    indice_programa = lista_programas.index(programa_elegido)
    indice_campus = lista_campus.index(campus)
    campus_elegido = valor_campus[indice_programa][indice_campus]
    if campus == 'london' and slots_disponible[campus_elegido] > 0:
        estudiante['campus'] = campus
        slots_disponible[campus_elegido] -= 1
        return estudiante
    elif campus == 'manchester' and slots_disponible[campus_elegido] > 0:
        estudiante['campus'] = campus
        slots_disponible[campus_elegido] -= 1
        return estudiante
    elif campus == 'liverpool' and slots_disponible[campus_elegido] > 0:
        estudiante['campus'] = campus
        slots_disponible[campus_elegido] -= 1
        return estudiante
    else:
        system('cls')
        print('No quedan slots disponible es este campus')
        return None


# Función para verificar si existen estudiantes
def verificar_estudiantes(indice_programa, indice_campus):
    lista = estudiantes[indice_programa][indice_campus]
    if len(lista) > 0:
        return True
    else:
        return False


# Función para selecciona menus
def mostrar_menus():
    while True:
        print('-' * 50)
        indice_menu = 1
        for programa in lista_programas:
            print(f'[{indice_menu}] {programa.capitalize()}')
            indice_menu += 1
        try:
            programa = input('Elige un programa: ')
            ['1', '2', '3', '4'].index(programa)
            system('cls')
            indice_campus = 1
            for campus in lista_campus:
                print(f'[{indice_campus}] {campus.capitalize()}')
                indice_campus += 1
            campus = input('Elige un campus: ')
            ['1', '2', '3'].index(campus)
        except ValueError:
            system('cls')
            print('Error, elige un programa o campus correcto')
        else:
            system('cls')
            indice_programa = int(programa) - 1
            indice_campus = int(campus) - 1
            ubicacion_estudiante = estudiantes[indice_programa][indice_campus]
            validacion = verificar_estudiantes(indice_programa, indice_campus)
            if validacion:
                for estudiante in ubicacion_estudiante:
                    system('cls')
                    for key, value in estudiante.items():
                        print(f'{key.capitalize()} : {value.capitalize()}')
                    input('Presione cualquier tecla para continuar')
                system('cls')
                return
            else:
                print('No existen estudiantes actualmente en este lugar')
                input('Presione cualquier tecla para continuar')
                system('cls')
                return


# Función del menu del programa
def menu():
    while True:
        print('Bienvenido al sistema de inscripción de Universidad')
        print('-' * 50)
        print('[1] Crear nuevo estudiante\n[2] Ver lista de estudiantes\n[3] Finalizar programa')
        try:
            accion = input('Elige una opción: ')
            ['1', '2', '3'].index(accion)
        except ValueError:
            system('cls')
            print('Error, escoge una opción correcta')
        else:
            system('cls')
            return accion


# Función principal
def main():
    system('cls')
    validacion = inicio_sesion()
    while validacion:
        accion = menu()
        if accion == '1':
            estudiante = crear_estudiante()
            elegir_campus(estudiante)
            continue
        elif accion == '2':
            mostrar_menus()
            continue
        else:
            system('cls')
            print('El programa ha sido finalizado')
            validacion = False


if __name__ == '__main__':
    main()
